app/models/service_account.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class ServiceAccount < User
```

Comment:
Should it have been inherited from `Principal? Could you elaborate why you choose User? I am just being curious.

Reply:
Principal was my first goal, because principal is "less" stuff.

In a nutshell the main conflicts were:

* We assign the thing (on the warden level) to an accessor called `current_user`, so even if it had worked, it would invite misconceptions and wrong ideas
* The concept of _permissions_ (which is relied upon heavily in APIv3) is added on the user-level, not on the Principal-level

In the end I made my peace with it as a subclass of user, because now it's on the same level (as in: depth of the inheritance tree) as the `SystemUser`, which seems to make sense to me.